test/jdk/java/io/ObjectInputStream/NegativeArraySizeTest.java

```diff
@@ -0,0 +1,112 @@
+        ObjectOutputStream oos = new ObjectOutputStream(baos);
+        oos.writeObject(simpleArray);
+        oos.close();
+        byte[] serializedData = baos.toByteArray();
+
+        // Find the right location to modify, looking for the first instance of TC_ENDBLOCKDATA
+        int firstPos = 0;
+        for (int i=0; i<serializedData.length-1; i++) {
+            // 0x78 = TC_ENDBLOCKDATA
+            if (serializedData[i] == 0x78) {
+                firstPos = i;
+                break;
+            }
+        }
+
+        // Replace array length with -2
+        serializedData[firstPos+2] = (byte) 0xff;
+        serializedData[firstPos+3] = (byte) 0xff;
+        serializedData[firstPos+4] = (byte) 0xff;
+        serializedData[firstPos+5] = (byte) 0xfe;
+
+        return  serializedData;
+    }
+
+    public static void test_simpleArray_negative() throws IOException, ClassNotFoundException {
+        ObjectInputStream ois = new ObjectInputStream(new ByteArrayInputStream(_buildPayload()));
+        ois.readObject();
+    }
+
+    public static void test_simpleArray_negative_with_filter() throws IOException, ClassNotFoundException {
+        class CustomFilter implements ObjectInputFilter {
+            @Override
+            public Status checkInput(FilterInfo filterInfo) {
+                Class<?> cl = filterInfo.serialClass();
+                if (cl != null && cl.isArray() && filterInfo.arrayLength() < -1) {
+                    throw new RuntimeException("FilterInfo.arrayLength() must be >= -1 for arrays (was " + filterInfo.arrayLength() + ")");
+                }
+                return Status.ALLOWED;
+            }
+        }
+        ObjectInputStream ois = new ObjectInputStream(new ByteArrayInputStream(_buildPayload()));
+        ois.setObjectInputFilter(new CustomFilter());
+        ois.readObject();
+    }
+
+    public static void main(String[] args) throws Exception {
+        try {
+            test_simpleArray_negative();
+        } catch (NegativeArraySizeException nase) {
+            throw new Exception("ObjectInputStream::readObject() shouldn't throw a NegativeArraySizeException", nase);
+        } catch (ObjectStreamException ose) {
+            // OK, because a NegativeArraySizeException should be converted into a ObjectStreamException
+        }
+        try {
+            test_simpleArray_negative_with_filter();
+        } catch (NegativeArraySizeException nase) {
+            throw new Exception("ObjectInputStream::readObject() shouldn't throw a NegativeArraySizeException", nase);
+        } catch (ObjectStreamException ose) {
+            if (ose instanceof InvalidClassException ice && ice.getMessage().contains("filter status: REJECTED")) {
+                throw new Exception("ObjectInputStream::readObject() should catch NegativeArraySizeExceptions before filtering", ice);
+            }
+            // OK, because a NegativeArraySizeException should be converted into a ObjectStreamException *before* filtering
```

Comment:
Sounds like we should really test for ICE with "Array length ... is negative" here. This would cover the case when filter rejected the class before that exception is thrown.

Reply:
The test for the `InvalidClassException` created by the object input filter is just above the comment, but I added another check for the correct exception message just for the simple test without filter.